psl-core/src/main/java/org/linqs/psl/reasoner/term/online/OnlineTermStore.java

```diff
@@ -99,6 +115,92 @@ public synchronized void updateLocalVariable(ObservedAtom atom, float newValue)
         variableValues[getVariableIndex(atom)] = newValue;
     }
 
+    public synchronized boolean activateRule(Rule rule) {
+        List<Integer> rulePages = rulePageMapping.get(rule);
+        if (rulePages == null) {
+            // Rule not in model.
+            return false;
+        }
+
+        activatedRules.put(rule, true);
+
+        int activePageIndex = 0;
+        for (Integer pageIndex : rulePages) {
+            activePageIndex = activeTermPages.indexOf(pageIndex);
+            if (activePageIndex == -1) {
+                activeTermPages.add(pageIndex);
+                numPages++;
+            }
+        }
+        return true;
+    }
+
+    public synchronized boolean addRule(Rule rule) {
+        List<Integer> rulePages = rulePageMapping.get(rule);
+        if (rulePages != null) {
+            // Rule already exists in model.
+            return false;
+        }
+
+        // Add new rule to rule set.
+        rules.add(rule);
+        rulePageMapping.put(rule, new ArrayList<Integer>());
+        activatedRules.put(rule, true);
+
+        // Ground new rule.
+        this.initialRound = true;
+        StreamingIterator<T> groundingIterator = getGroundingIterator(Arrays.asList(rule));
+        while (groundingIterator.hasNext()) {
+            groundingIterator.next();
+        }
+
+        return true;
+    }
+
+    public synchronized boolean deactivateRule(Rule rule) {
+        List<Integer> rulePages = rulePageMapping.get(rule);
+        if (rulePages == null) {
+            // Rule does not exist in model.
+            return false;
+        }
+
+        activatedRules.put(rule, false);
+
+        int activePageIndex = 0;
+        for (Integer i : rulePages) {
```

Comment:
Use `index` of `pageIndex` (like you have elsewhere in this file) instead of `i` here.
`i` is generally for incrementers, here we are iterating through arbitrary indexes.